theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import absolute_import, print_function, division
+
+import numpy as np
+import theano
+from theano import Op
+from theano import config
+import theano.tensor as T
+from theano.tensor.extra_ops import cpu_contiguous
+from .basic_ops import (gpu_contiguous, as_gpuarray_variable,
+                        infer_context_name, CGpuKernelBase)
+import theano.tensor.nnet.ctc
+from .type import GpuArrayType
+from .opt import register_opt, op_lifter, register_opt2
+from theano.gradient import grad_undefined
+from theano import gof
+from theano.gof import local_optimizer
+from theano.tensor.opt import register_canonicalize
+from theano.tensor.opt import register_stabilize
+
+import os
+import pygpu
+
+ctc_enabled = config.ctc.enabled
+
+
+class GpuConnectionistTemporalClassification(CGpuKernelBase, Op):
+    """
+    GPU wrapper for Baidu CTC loss function.
+    """
+    __props__ = ('context_name', 'compute_grad',)
+
+    func_file = "./ctc_wrapper.c"
+    func_name = "APPLY_SPECIFIC(ctc_cost_gpu)"
+
+    def __init__(self, compute_grad=True, context_name=None):
+        if not compute_grad:
+            self.func_name = "APPLY_SPECIFIC(ctc_cost_gpu_no_grad)"
+        self.compute_grad = compute_grad
+        self.context_name = context_name
+
+        Op.__init__(self)
+        CGpuKernelBase.__init__(self, self.func_file, self.func_name)
+
+        self.costs = GpuArrayType(dtype='float32',
+                                  broadcastable=(False,),
+                                  context_name=self.context_name)
+
+        if self.compute_grad:
+            self.gradients = GpuArrayType(dtype='float32',
+                                          broadcastable=(False, False, False,),
+                                          context_name=self.context_name)
+
+        if config.ctc.root == "":
+            raise ValueError('ctc.root variable is not set, please set it '
+                             'to the root directory of the CTC library in '
+                             'your system.')
+
+    def c_lib_dirs(self):
+        dirs = []
+        if ctc_enabled:
+            # We assume here that the compiled library (libwarpctc.so) is available
+            # at the build directory of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "build"))
+        return dirs
+
+    def c_libraries(self):
+        return ["warpctc", "gpuarray"]
+
+    def c_header_dirs(self):
+        dirs = [os.path.dirname(__file__), pygpu.get_include()]
+        if ctc_enabled:
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        dirs.append('/usr/local/cuda/include')
```

Comment:
Baidu CTC library does not currently support Windows, only Mac OS X and Linux. But if the approach using only PyGpu is utilized, than the cuda path is no longer required.